data_steward/retraction/retract_deactivated_pids.py

```diff
@@ -0,0 +1,498 @@
+FROM `{{deactivated_pids_project}}.{{deactivated_pids_dataset}}.{{deactivated_pids_table}}`
+WHERE person_id = {{pid}} ) ELSE CASE WHEN {{end_date_column}} IS NULL THEN {{start_date_column}} < (
+SELECT deactivated_date
+FROM `{{deactivated_pids_project}}.{{deactivated_pids_dataset}}.{{deactivated_pids_table}}`
+WHERE person_id = {{pid}}) END END)
+""")
+# Deactivated participant table fields to query off of
+PID_TABLE_FIELDS = [{
+    "type": "integer",
+    "name": "person_id",
+    "mode": "required",
+    "description": "The person_id to retract data for"
+}, {
+    "type": "date",
+    "name": "deactivated_date",
+    "mode": "required",
+    "description": "The deactivation date to base retractions on"
+}]
+
+
+def add_console_logging(add_handler):
+    """
+    This config should be done in a separate module, but that can wait
+    until later.  Useful for debugging.
+    """
+    try:
+        os.makedirs(LOGS_PATH)
+    except OSError:
+        # directory already exists.  move on.
+        pass
+
+    name = datetime.now().strftime(
+        os.path.join(LOGS_PATH, 'ehr_deactivated_retraction-%Y-%m-%d.log'))
+    logging.basicConfig(filename=name,
+                        level=logging.INFO,
+                        format='{asctime} - {name} - {levelname} - {message}',
+                        style='{')
+
+    if add_handler:
+        handler = logging.StreamHandler()
+        handler.setLevel(logging.INFO)
+        formatter = logging.Formatter('{levelname} - {name} - {message}',
+                                      style='{')
+        handler.setFormatter(formatter)
+        logging.getLogger('').addHandler(handler)
+
+
+def get_pids_table_info(project_id, dataset_id):
+    """
+    This function gets all table information for the dataset and filters to only retain tables with the field person_id
+
+    :param project_id: bq name of project_id
+    :param dataset_id: ba name of dataset_id
+    :return: dataframe with table information schema from BigQuery
+    """
+    all_table_info_query = TABLE_INFORMATION_SCHEMA.render(project=project_id,
+                                                           dataset=dataset_id)
+    result_df = bq.query(q=all_table_info_query, project_id=project_id)
+    # Get list of tables that contain person_id
+    pids_tables = []
+    for i, row in result_df.iterrows():
+        column = getattr(row, 'column_name')
+        table = getattr(row, 'table_name')
+        if 'person_id' in column:
+            pids_tables.append(table)
+    pids_tables_info_df = result_df[result_df['table_name'].isin(pids_tables)]
+    return pids_tables_info_df
+
+
+def get_date_info_for_pids_tables(project_id):
+    """
+    Loop through tables within all datasets and determine if the table has an end_date date or a date field. Filtering
+    out person and death table and keeping only tables with PID and an upload or start/end date associated.
```

Comment:
This is a good question- I am omitting the person table since the person could still have records before deactivation date, which would in turn need to reference the person table. I then assumed to leave the death table. But I could be completely wrong here. Let me reach out to Caroline.

Reply:
My assumption was wrong, updating this now.